src/log/main.cr

```diff
@@ -0,0 +1,142 @@
+class Log
+  # Creates a `Log` for the given source.
+  # If *level* is given, it will override the configuration.
+  def self.for(source : String, level : Severity? = nil) : Log
+    log = builder.for(source)
+    log.level = level if level
+    log
+  end
+
+  # Creates a `Log` for the given nested source.
+  # If *level* is given, it will override the configuration.
+  def for(child_source : String, level : Severity? = nil) : Log
+    return ::Log.for(child_source) if source.blank?
+    return ::Log.for(source) if child_source.blank?
+    ::Log.for("#{source}.#{child_source}", level)
+  end
+
+  # Creates a `Log` for the given type.
+  # A type `Foo::Bar(Baz)` corresponds to the source `foo.bar`.
+  # If *level* is given, it will override the configuration.
+  def self.for(type : Class, level : Severity? = nil) : Log
+    source = type.name.underscore.gsub("::", ".")
+
+    # remove generic arguments
+    paren = source.index('(')
+    source = source[0...paren] if paren
+
+    ::Log.for(source, level)
+  end
+
+  # :ditto:
+  def for(type : Class, level : Severity? = nil) : Log
+    ::Log.for(type, level)
+  end
+
+  private Top = Log.for("")
+
+  {% for method in %i(debug verbose info warn error fatal) %}
+    # See `Log#{{method.id}}`.
+    def self.{{method.id}}(*, exception : Exception? = nil)
+      Top.{{method.id}}(exception: exception) do
+        yield
+      end
+    end
+  {% end %}
+
+  @@builder = Builder.new
+
+  # Returns the default `Log::Builder` used for `Log.for` calls.
+  def self.builder
+    @@builder
+  end
+
+  # Returns the current fiber logging context.
+  def self.context : Log::Context
+    Fiber.current.logging_context
+  end
+
+  # Sets the current fiber logging context.
+  def self.context=(value : Log::Context)
+    Fiber.current.logging_context = value
+  end
+
+  # Returns the current fiber logging context.
+  def context : Log::Context
+    Log.context
+  end
+
+  # Sets the current fiber logging context.
+  def context=(value : Log::Context)
+    Log.context = value
+  end
+
+  class Context
```

Comment:
Why is the `Context` class split between two files? It's a little confusing...

Reply:
A matter of taste. The `log/main.cr` describes the api the user will use. I found that more clear to have a better perspective and express the intention of the methods.